tests/activity.go

```diff
@@ -988,3 +989,135 @@ func (s *ClientFunctionalSuite) TestActivityHeartbeatDetailsDuringRetry() {
+					ActivityId:             convert.IntToString(rand.Int()),
+					ActivityType:           &commonpb.ActivityType{Name: activityName},
+					TaskQueue:              taskQueue,
+					Input:                  payloads.EncodeString("activity-input"),
+					Header:                 header,
+					ScheduleToCloseTimeout: durationpb.New(15 * time.Second),
+					ScheduleToStartTimeout: durationpb.New(60 * time.Second),
+					StartToCloseTimeout:    durationpb.New(15 * time.Second),
+					HeartbeatTimeout:       durationpb.New(60 * time.Second),
+				}},
+			}}, nil
+		case enumspb.COMMAND_TYPE_COMPLETE_WORKFLOW_EXECUTION:
+			workflowComplete = true
+			return []*commandpb.Command{{
+				CommandType: enumspb.COMMAND_TYPE_COMPLETE_WORKFLOW_EXECUTION,
+				Attributes: &commandpb.Command_CompleteWorkflowExecutionCommandAttributes{CompleteWorkflowExecutionCommandAttributes: &commandpb.CompleteWorkflowExecutionCommandAttributes{
+					Result: payloads.EncodeString("Done"),
+				}},
+			}}, nil
+		}
+		panic("Unexpected workflow state")
+	}
+
+	heartbeatSignalChan := make(chan bool) // Activity task sends heartbeat when signaled on this chan. It also signals back on the same chan after sending the heartbeat.
+	endActivityTask := make(chan bool)     // Activity task completes when signaled on this chan. This is to force the task to be in pending state.
+	atHandler := func(
+		execution *commonpb.WorkflowExecution, activityType *commonpb.ActivityType, activityID string, input *commonpb.Payloads, taskToken []byte,
+	) (*commonpb.Payloads, bool, error) {
+
+		<-heartbeatSignalChan // wait for signal before sending heartbeat.
+		_, err := s.engine.RecordActivityTaskHeartbeat(NewContext(), &workflowservice.RecordActivityTaskHeartbeatRequest{
+			Namespace: s.namespace,
+			TaskToken: taskToken,
+			Details:   payloads.EncodeString("details"),
+			Identity:  workerIdentity, // explicitly set the worker identity in the heartbeat request
+		})
+		s.NoError(err)
+		heartbeatSignalChan <- true // signal that the heartbeat is sent.
+
+		<-endActivityTask // wait for signal before completing the task
+		return payloads.EncodeString("Activity Result"), false, nil
+	}
+
+	poller := &TaskPoller{
+		Engine:              s.engine,
+		Namespace:           s.namespace,
+		TaskQueue:           taskQueue,
+		Identity:            "", // Do not send the worker identity.
+		WorkflowTaskHandler: wtHandler,
+		ActivityTaskHandler: atHandler,
+		Logger:              s.Logger,
+		T:                   s.T(),
+	}
+
+	// execute workflow task so that an activity can be enqueued.
+	_, err = poller.PollAndProcessWorkflowTask()
+	s.True(err == nil || err == errNoTasks)
+
+	// execute activity task which waits for signal before sending heartbeat.
+	go func() {
+		err := poller.PollAndProcessActivityTask(false)
+		s.True(err == nil || err == errNoTasks)
+	}()
+
+	heartbeatSignalChan <- true // ask the activity to send a heartbeat.
```

Comment:
I would verify after this line that the workerIdentity from DescribeWorkflowExecution is empty here

Reply:
Done.